lib/src/core/networking/client.dart

```diff
@@ -538,4 +491,62 @@ abstract class OpenAINetworkingClient {
   static bool doesErrorExists(Map<String, dynamic> decodedResponseBody) {
     return decodedResponseBody[OpenAIStrings.errorFieldKey] != null;
   }
+
+  static RequestFailedException extractException(int statusCode, Map<String, dynamic> errorBody, Map<String, String> headers) {
+    final message = errorBody[OpenAIStrings.messageFieldKey];
+    final msg = message?.toString() ?? "";
+
+    if (statusCode == _rateLimitStatusCode) {
+      final remainingValue = headers[_headerRequestsRemaining];
+      final remaining = int.tryParse(remainingValue ?? "");
+      final resetValue = headers[_headerResetIn];
+      if (resetValue != null && resetValue.isNotEmpty) {
+        final timeoutDuration = parseResetTimeout(resetValue);
+        final timedOutUntil = DateTime.now().add(timeoutDuration);
+
+        return RateLimitedException(msg, statusCode, timeoutDuration, timedOutUntil);
+      } else if (remaining != null && remaining <= 0) {
+        return OutOfCreditsException(msg, statusCode);
```

Comment:
Cool, I really like your idea here,  am thinking of leveraging this to support more customized exceptions and document it as well for better developer experience.